lambda/matching.ml

```diff
@@ -1232,12 +1232,23 @@ and precompile_var args cls def k =
                 | PmOr { or_matrix = m } -> m
                 | PmVar x -> add_omega_column (rebuild_matrix x.inside)
               in
-              let rec rebuild_default nexts def =
-                match nexts with
-                | [] -> def
-                | (e, pmh) :: rem ->
-                    (add_omega_column (rebuild_matrix pmh), e)
-                    :: rebuild_default rem def
+              let rebuild_default nexts def =
+                (* Assuming we had bound [top_default] from the result of
+                   [split_or], then we could have written this as:
+                   {[
+                     List.map
+                       (fun (mat, e) -> add_omega_column mat, e)
+                       top_default
+                   ]}
+                   But that would actually be doing more work: it would add an
+                   omega column to all the matrices in top_default, whereas here
+                   we're reusing the matrices that were in [def], and only
+                   adding the omega column to the new matrices.
```

Comment:
As a counter-point: `rebuild_matrix` does redo some work that has already been done in the recursive call.

It's not clear to me which version is more efficient, most likely it doesn't really matter, and perhaps we want to just use the alternative version shown in this comment (assuming it is actually equivalent), just because it is simpler to reason about, and more regular with what is done in `split_constr` and `split_naive`.

Reply:
After thinking quite a bit about the whole thing (these defaults are tricky things, maybe I should re-read the paper after all!), I don't think that the comment is correct. It seems to assume that popping the default columns (as done in `let var_def = default_pop_column def`) and then adding omega columns (as done on the tail part of that `top_default` that is `var_def`) are inverse operations. But that is only true if the column popped where omega-columns in the first place, and there is no reason for this to hold. In the general case (the first column of some default matrices are not omega), then the suggested simpler implementation would lose information.

Reply:
> But that is only true if the column popped where omega-columns in the first place, and there is no reason for this to hold.

Indeed!
I will update the comment to reflect that.